content/languages/cpp/authoring/index.md

```diff
@@ -0,0 +1,369 @@
+
+```cpp
+
+#include <random>
+
+Describe(TestSuite) {
+private:
+
+  //One random engine is usually enough.
+  std::mt19937 engine{ std::random_device{}() };
+
+  //A set of callables for every type or range you are going to need in your tests.
+  //For convenience, distributions are already bound with the PRNG.
+  std::function<int   ()> rand_number      = std::bind(std::uniform_int_distribution<    int>{  1,  100 }, engine);
+  std::function<size_t()> rand_length      = std::bind(std::uniform_int_distribution< size_t>{ 20,  100 }, engine);
+  std::function<double()> rand_coefficient = std::bind(std::uniform_real_distribution<double>{ -1,    1 }, engine);
+  std::function<char  ()> rand_letter      = std::bind(std::uniform_int_distribution<   char>{ 'a', 'z' }, engine);
+  std::function<bool  ()> rand_bool        = std::bind(std::uniform_int_distribution<   char>{   0,   1 }, engine);
+  // and etc.
+
+public:
+  It(RandomTests) {
+
+    size_t input_length = rand_length(); //use length generator
+    
+    std::string input;
+    for(int i=0; i<input_length; ++i) {
+      
+      //string generation logic...
+      input.push_back(rand_letter()); //use generator of random letters
+
+    }
+
+    //... testing logic    
+  }
+};
+```
+
+_TODO: `sample`, `shuffle`
+
+### Reference solution
+
+If the test suite happens to use a reference solution to calculate expected values (which [should be avoided](/authoring/guidelines/submission-tests/#reference-solution) when possible), or some kind of reference data like precalculated arrays, etc., it must not be possible for the user to call it, redefine, overwrite, or directly access its contents. To prevent this, reference data can be made a private member of `Describe` structure. Reference solution can be either a private member of `Describe`, or a lambda-initialized local variable of an `It` block. See `[Example test suite](#example_test_suite)` for some examples.
+
+The reference solution or data ___must not___ be defined in the [Preloaded code](/authoring/guidelines/preloaded/).
+
+
+### Input mutation
+
+General guidelines for submission tests contain a section related to [input mutation](/authoring/guidelines/submission-tests/#input-mutation) and how to prevent users from abusing it to work around kata requirements. It's usually not a problem in "real world" C++ programming, but on Codewars, users can take advantage of vulnerable test suites and modify their behavior. Constness of a function argument can be forcefully cast away by a user, or they can change the function signature and remove `const` qualifier from input parameter(s).  After calling a user solution, tests should not rely on the state of such values and they should consider them as potentially modified by a user.
```

Comment:
```suggestion
General guidelines for submission tests contain a section related to [input mutation](/authoring/guidelines/submission-tests/#input-mutation) and how to prevent users from abusing it to work around kata requirements. It's usually not a problem in "real world" C++ programming, but on Codewars, users can take advantage of vulnerable test suites and modify their behavior. Constness of a function argument can be forcefully cast away by a user, or they can change the function signature and remove `const` qualifier from input parameter(s).  After calling a user solution, tests should not rely on the state of such values, and they should consider them as potentially modified by a user.
```